src/dbnode/storage/shard.go

```diff
@@ -1897,7 +1891,254 @@ func (s *dbShard) Flush(
+			BlockStart:        startTime,
+			DeleteIfExists:    false,
+		}
+		prepared, err := flushPreparer.PrepareData(prepareOpts)
+		if err != nil {
+			s.incrementFlushStateFailures(startTime)
+			multiErr = multiErr.Add(err)
+			continue BlockLoop
+		}
+
+		// Persistence is done in two stages. The first stage is to loop through
+		// series on disk and merge it with what's in memory. The second stage
+		// is to persist the rest of the series in memory that was not
+		// persisted in the first stage.
+
+		// First stage: loop through series on disk.
+		for id, tagsIter, data, _, err := fsReader.Read(); err != io.EOF; id, _, data, _, err = fsReader.Read() {
+			if err != nil {
+				s.incrementFlushStateFailures(startTime)
+				multiErr = multiErr.Add(err)
+				continue BlockLoop
+			}
+
+			// There will be at most two BlockReader slices: one for disk data
+			// and one for in memory data.
+			brs := make([][]xio.BlockReader, 0, 2)
+
+			// Create BlockReader slice out of disk data.
+			seg := ts.NewSegment(data, nil, ts.FinalizeHead)
+			sr := srPool.Get()
+			sr.Reset(seg)
+			br := xio.BlockReader{
+				SegmentReader: sr,
+				Start:         startTime,
+				BlockSize:     blockSize,
+			}
+			brs = append(brs, []xio.BlockReader{br})
+
+			// Check if this series is in memory (and thus requires merging).
+			if element, ok := dirtySeries.Get(idAndBlockStart{blockStart: blockStart, id: id}); ok {
+				// Series is in memory, so it will get merged with disk and
+				// written in this loop. Therefore, we need to remove it from
+				// the "to write" list so that the later loop does not rewrite
+				// it.
+				dirtySeriesToWrite[blockStart].Remove(element)
+
+				s.RLock()
+				entry, _, err := s.lookupEntryWithLock(element.Value)
+				s.RUnlock()
+
+				tmpCtx.Reset()
+				encoded, err := entry.Series.FetchBlocksForColdFlush(tmpCtx, startTime, nextVersion)
+				tmpCtx.BlockingClose()
+				if err != nil {
+					s.incrementFlushStateFailures(startTime)
+					multiErr = multiErr.Add(err)
+					continue BlockLoop
+				}
+
+				if len(encoded) > 0 {
+					brs = append(brs, encoded)
+				}
+			}
+
+			mergedIter := multiIterPool.Get()
+			mergedIter.ResetSliceOfSlices(xio.NewReaderSliceOfSlicesFromBlockReadersIterator(brs))
+			defer mergedIter.Close()
+
+			tags, err := convert.TagsFromTagsIter(id, tagsIter, s.opts.IdentifierPool())
+			if err != nil {
+				s.incrementFlushStateFailures(startTime)
+				multiErr = multiErr.Add(err)
+				continue BlockLoop
+			}
+			if err := persistIter(prepared.Persist, id, tags, mergedIter, encoderPool); err != nil {
+				s.incrementFlushStateFailures(startTime)
+				multiErr = multiErr.Add(err)
+				continue BlockLoop
+			}
+		}
+
+		// Second stage: loop through rest of series in memory that was not on
+		// disk.
+		for seriesElement := seriesList.Front(); seriesElement != nil; seriesElement = seriesElement.Next() {
+			s.RLock()
+			entry, _, err := s.lookupEntryWithLock(seriesElement.Value)
```

Comment:
Is this safe? I guess the series could not have been evicted in the meantime

Reply:
Yeah I don't think it's possible that the entry would become invalid within the lifetime of this function.